### PR TITLE
Add a simple fuzz test for jsoncpp.

### PR DESCRIPTION
**Pull request by Google-Autofuzz**
This pull-request is an update of #610

### CODE REVIEW

File: src/test_lib_json/fuzz.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2007-2010 The JsonCpp Authors
+// Distributed under MIT license, or public domain if desired and
+// recognized in your jurisdiction.
+// See file LICENSE for detail or copy at http://jsoncpp.sourceforge.net/LICENSE
+
+#include "fuzz.h"
+
+#include <bits/stdint-uintn.h>
```
Comment (baylesj) [inline src/test_lib_json/fuzz.cpp]: Non-standard header is what is breaking the build. Replace with:

```
#include <cstdint>
```

File: src/test_lib_json/fuzz.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2007-2010 The JsonCpp Authors
```
Comment (baylesj) [inline src/test_lib_json/fuzz.cpp]: Update header?

File: src/test_lib_json/fuzz.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  Json::CharReaderBuilder builder;
+
+  if (size < sizeof(uint32_t)) {
+    return 0;
+  }
+
+  uint32_t hash_settings = *(const uint32_t*)data;
```
Comment (baylesj) [inline src/test_lib_json/fuzz.cpp]: Nit: const

File: src/test_lib_json/fuzz.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2007-2010 The JsonCpp Authors
+// Distributed under MIT license, or public domain if desired and
+// recognized in your jurisdiction.
+// See file LICENSE for detail or copy at http://jsoncpp.sourceforge.net/LICENSE
+
+#ifndef FUZZ_H_INCLUDED
+#define FUZZ_H_INCLUDED
+
+#include <stddef.h>
```
Comment (baylesj) [inline src/test_lib_json/fuzz.h]: Nit: prefer C++ headers (cstddef over stddef.h)

File: src/test_lib_json/main.cpp [hunk 5]
```diff
@@ -2607,6 +2620,8 @@ int main(int argc, const char* argv[]) {
 
   JSONTEST_REGISTER_FIXTURE(runner, RValueTest, moveConstruction);
```
Comment (baylesj) [inline src/test_lib_json/main.cpp]: Nit: remove extra newline.

File: src/test_lib_json/fuzz.h [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+#include <stdint.h>
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size);
```
Comment (baylesj) [inline src/test_lib_json/fuzz.h]: Do we actually need extern "C"? I thought this was all C++ code.

File: src/test_lib_json/main.cpp (hunk 5 above)
Comment (Google-Autofuzz) [reply src/test_lib_json/main.cpp]: Are you referring to the newline before or after line 2621? Or are you referring to line 2623?

File: src/test_lib_json/fuzz.h (hunk 4 above)
Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.h]: Done.

File: src/test_lib_json/fuzz.cpp (hunk 2 above)
Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.cpp]: Done.

File: src/test_lib_json/fuzz.cpp (hunk 1 above)
Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.cpp]: Done.

File: src/test_lib_json/fuzz.cpp (hunk 3 above)
Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.cpp]: Done.

File: src/test_lib_json/fuzz.h (hunk 6 above)
Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.h]: Yes, we do need it so that it can find LLVM.

File: src/test_lib_json/main.cpp (hunk 5 above)
Comment (baylesj) [reply src/test_lib_json/main.cpp]: I think this is fine as-is actually, the fixture spacing is just excessive.

File: src/test_lib_json/fuzz.h (hunk 6 above)
Comment (baylesj) [reply src/test_lib_json/fuzz.h]: This is still failing to get linked properly on some platforms. I've updated the build bots to be more modern, which might help remove this issue. If we fix the compile problem I think this patch is good to go.

```
jsoncpp_test@exe/src_test_lib_json_main.cpp.o: In function `TestFuzzTestfuzzDoesntCrash::runTestCase()':
../src/test_lib_json/main.cpp:(.text+0x2c08a): undefined reference to `LLVMFuzzerTestOneInput'
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.h]: Do you know how I can fix this compile problem? I couldn't find the bug for this error.

Comment (baylesj) [reply src/test_lib_json/fuzz.h]: Not sure, but I do see it locally building with meson + ninja:

```
FAILED: jsoncpp_test 
c++  -o jsoncpp_test 'jsoncpp_test@exe/src_test_lib_json_jsontest.cpp.o' 'jsoncpp_test@exe/src_test_lib_json_main.cpp.o' -Wl,-dead_strip_dylibs libjsoncpp.21.dylib -Wl,-headerpad_max_install_names -Wl,-rpath,@loader_path/
Undefined symbols for architecture x86_64:
  "_LLVMFuzzerTestOneInput", referenced from:
      TestFuzzTestfuzzDoesntCrash::runTestCase() in src_test_lib_json_main.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Comment (baylesj) [reply src/test_lib_json/fuzz.h]: Figured it out. We don't build with CMake--that's a nicety for some of our clients. Internally we use the meson+ninja build outlined in the docs, which require updating a different build file, the meson.build. The test compiles successfully with the below change:

```
-> % git diff
diff --git a/meson.build b/meson.build
index ece8aa4..8de947c 100644
--- a/meson.build
+++ b/meson.build
@@ -88,7 +88,8 @@ jsoncpp_test = executable(
   'jsoncpp_test',
   [ 'src/test_lib_json/jsontest.cpp',
     'src/test_lib_json/jsontest.h',
-    'src/test_lib_json/main.cpp'],
+    'src/test_lib_json/main.cpp',
+    'src/test_lib_json/fuzz.cpp'],
   include_directories : jsoncpp_include_directories,
   link_with : jsoncpp_lib,
   install : false,
```

Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.h]: Thanks! It passed!

Comment (Google-Autofuzz) [reply src/test_lib_json/fuzz.h]: It can be merged.

**Pull request merged**